src/dmd/expressionsem.d

```diff
@@ -9624,15 +9624,15 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 exp.e2 = exp.e2.implicitCastTo(sc, tb1next);
                 exp.type = tb1next.arrayOf();
             L2elem:
-                if (tb2.ty == Tarray || tb2.ty == Tsarray)
+                if (tb2.ty != Tarray && tb2.ty != Tsarray)
                 {
                     // Make e2 into [e2]
                     exp.e2 = new ArrayLiteralExp(exp.e2.loc, exp.type, exp.e2);
                 }
                 else if (checkNewEscape(sc, exp.e2, false))
                     return setError();
-                result = exp.optimize(WANTvalue);
-                return;
+                e = exp.optimize(WANTvalue);
+                goto Lrewrite;
```

Comment:
Can you place the rewriting into a (local) function instead of having to use goto?